### PR TITLE
Adding versioning support to Cerberus.

### PR DESCRIPTION
**Pull request by splintercat**
Decided to add support for secret versioning.

I tried to add some more test coverage.

Let me know if there are style things, or really anything else you'd like changed.

### CODE REVIEW

File: cerberus/client.py [hunk 1]
```diff
@@ -291,13 +317,14 @@ def update_sdb(self, sdb_id, owner=None, description=None, user_group_permission
 
     def delete_secret(self, vault_path):
         """Delete a secret from the given vault path"""
-        secret_resp = requests.delete(self.cerberus_url + '/v1/secret/' + vault_path,
+        secret_resp = requests.delete(self.cerberus_url + '/v1/secret/' + str.strip(vault_path, '/'),
```
[the remark this answers is not in the document]
Comment (splintercat) [reply cerberus/client.py]: But they never actually have a '/' at the end (or beginning) of a path, which is what that strip is taking care of.  I put in the same logic elsewhere, where there were situations where someone could mistakenly leave a trailing '/', which would cause an error.

In the case of deletion, with a key in a virtual path, you cannot delete the virtual path directly.  If you delete all of the keys underneath a virtual path, then the virtual path is automatically deleted as well.

so 
```python
client.put_secret('category/sdb/virt/key', {'key': 'value in virtual path'})
client.list_secret('category/sdb/')
 # returns something like
 # {'......
 #   'data': {'keys': ['virt/']},
 # ......}
client.list_secret('category/sdb/virt/')
# Returns
 # {'......
 #   'data': {'keys': ['key']},
 # ......}

# this is the behavior before adding the stripping
client.delete_secret('category/sdb/virt/')
# ERROR

client.delete_secret('category/sdb/virt')
# ERROR

client.delete_secret('category/sdb/virt/key')
# Success

client.list_secret('category/sdb/')
 # returns something like
 # {'......
 #   'data': {'keys': []},
 # ......}
```

What the strip command helps prevent is when someone issues a
```python
client.delete_secret('category/sdb/virt/key/')
```
Which originally will cause an error, will just delete the key properly after the change.


Now with that all said..  I'm OK with removing the strip command from delete_secret, since someone is much more likely to be more specific when they are deleting a key.

File: cerberus/client.py [hunk 2]
```diff
@@ -327,30 +372,69 @@ def get_secrets(self, vault_path):
...
-        return secret_resp.json()['data']
+        if(offset is None):
+            offset = 0
 
+        secret_resp = requests.get(str.join('', [self.cerberus_url, '/v1/secret-versions/', vault_path,
+                                   '?limit=', str(limit), '&offset=', str(offset)]), headers=self.HEADERS)
+        self.throw_if_bad_response(secret_resp)
+        return secret_resp.json()
+
+    def get_all_secret_version_ids(self, vault_path, limit=None):
```
Comment (splintercat) [inline cerberus/client.py]: For both get_all_secret_version_ids, and get_all_secret_version_ids.

I'm going back and forth on whether I should include these functions, and even if they should be kept, if they are doing things the right way.

On one hand, a generator is a good way to handle something with the possibility of returning a ton of data.  But all of the other functions in the library don't use generators, so it seems like a divergence from the rest of the style of the library.

I'm also considering whether these functions should be made "private",  until others decide if they are truly useful or not.

In the same vein, I'm wondering if something like a contrib library would be a better place to put some convenience functions until a decision is made that they should be part of the core library.

File: cerberus/client.py [hunk 3]
```diff
@@ -327,30 +372,69 @@ def get_secrets(self, vault_path):
...
+                yield summary
+
+    def get_all_secret_versions(self, vault_path, limit=None):
+        """
+        Convience function that returns a generator yielding the contents of secrets and their version info
+        vault_path -- full path to the key in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        """
+        for secret in self.get_all_secret_version_ids(vault_path, limit):
+            yield {'secret': self.get_secrets_data(vault_path, version=secret['id']),
```
Comment (splintercat) [inline cerberus/client.py]: While convenient to package the secret and it's version info into a dictionary, I wonder if I'm being too presumptuous about what potential users of this library will actually want.

File: cerberus/client.py [hunk 4]
```diff
@@ -327,30 +372,69 @@ def get_secrets(self, vault_path):
...
+    def get_secret_versions(self, vault_path, limit=None, offset=None):
+        """
+        Get verions of a particular secret key
+
+        vault_path -- full path to the key in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        offset -- Default(0), used for pagination.  Will request records from the given offset.
+        """
+        # Set the normal defaults
+        if(limit is None):
```
Comment (dogonthehorizon) [inline cerberus/client.py]: \<nitpick\>
Something a little more Pythonic could be:

```python
if not limit:
    limit = 100
```

\</nitpick\>

File: cerberus/client.py [hunk 5]
```diff
@@ -327,30 +372,69 @@ def get_secrets(self, vault_path):
...
+            limit = 100
 
-        return secret_resp.json()['data']
+        if(offset is None):
+            offset = 0
 
+        secret_resp = requests.get(str.join('', [self.cerberus_url, '/v1/secret-versions/', vault_path,
```
Comment (dogonthehorizon) [inline cerberus/client.py]: [`requests` provides kwargs for adding parameters to URLs, you don't need to do this yourself](http://docs.python-requests.org/en/master/user/quickstart/#passing-parameters-in-urls)

File: cerberus/client.py [hunk 6]
```diff
@@ -327,30 +372,69 @@ def get_secrets(self, vault_path):
...
+
+    def get_all_secret_version_ids(self, vault_path, limit=None):
+        """
+        Convience function that returns a generator that will paginate over the secret version ids
+        vault_path -- full path to the key in the safety deposit box
+        limit -- Default(100), limits how many records to be returned from the api at once.
+        """
+        offset = 0
+        versions = self.get_secret_versions(vault_path, limit, offset)
+        for summary in versions['secure_data_version_summaries']:
```
Comment (dogonthehorizon) [inline cerberus/client.py]: I'd almost suggest having this in a smaller internal helper since we're calling it in both places. It'd be fairly easy to test that way too

File: tests/test_cerberus_client.py [hunk 7]
```diff
@@ -99,6 +99,72 @@ def test_get_token(self):
...
+        },
+        {
+            "id": "a7192aa7-83f0-45b7-91fb-f6b0eb",
+            "name": "snowflake",
+            "path": "app/snowflake/",
+            "category_id": "b042d0-e6-90-0aec03"
+        }
+    ]
+
+    #@patch('cerberus.client.CerberusClient.get_sdbs', return_value=sdb_data)
```
Comment (dogonthehorizon) [inline tests/test_cerberus_client.py]: Is this meant to stay in the codebase?

File: cerberus/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (splintercat) [reply cerberus/client.py]: I went back and did some testing both pre and post change and it didn't seem to negatively affect the process of deletion.

Though I highly encourage others to test that.  I may be missing a corner case.

File: cerberus/client.py (hunk 2 above)
Comment (dogonthehorizon) [reply cerberus/client.py]: Agreed, private is probably a good idea so that consumers don't depend on this functionality. I like the idea of generators, but I'm generally a fan of lazyness as a concept so 🤷‍♂️

File: cerberus/client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (splintercat) [reply cerberus/client.py]: Hmm I found an fascinating corner case.
With the code as it is.  You can create a blank secret.  Because I don't do any stripping on put_secret, you can do things like 
```python
client.put_secret('category/sdb/virt/', {'key': "The path to this secret is '' "})
```

So before I put in the stuff to do stripping, you could at still access the secret by leaving a trailing '/'.  Since I do the stripping for '/' in get_secrets, list_secrets, and delete_secret you can't actually access the blank key anymore.

While I'd love to prevent the creation of blank secrets, I feel like because the API and other libraries allow the creation of blank keys, I don't think I should be introducing a regression in features.

File: cerberus/client.py (hunk 6 above)
Comment (splintercat) [reply cerberus/client.py]: I've now condensed this down.

**Comment by sdford** [comment]
Now that this is approved, you mind bumping the version?

**Comment by splintercat** [comment]
Sure.  Should I update the Changelog as well?

**Comment by sdford** [comment]
Sure. That would be great.

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,9 @@ All notable changes to `Cerberus Python Client` will be documented in this file.
 
 ### Added
 
+###### 04-04-2018
+- Version 0.12.0 - Added support for versioning of secrets in Cerberus.
+
 ###### 02-06-2018
```
Comment (splintercat) [inline CHANGELOG.md]: Is this date correct?  Or should it be 06-02-2018.

Would it help if the Changelog specified the date format?

Comment (sdford) [reply CHANGELOG.md]: You're right it should be 06-02-2018.

Feel free to specify the format if you like.

**Comment by splintercat** [comment]
Thanks!
